### PR TITLE
[PATCH RFC] audio_stream: Introduce new API to copy data to/from raw buffer

### PR DESCRIPTION
**Pull request by dbaluta**
Introduce two new API functions, that will copy audio streams data
to/from raw buffers.

This is useful for the new codec adapter where we need to interface
data copying between SOF modules and library buffers.

For example, in this new usecase data sent from an audio component to
a decoder component is compressed so we cannot interpret it with builtin
types (s16/s24/s32) but we interpret it as 'raw'.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

### CODE REVIEW

**Review by paulstelian97 — COMMENTED** [review]
Shouldn't the function also check that there is data (for the `from_` variant) or room (for the `to_` variant) for the data, and limit the amount of data copied?

**Comment by dbaluta** [comment]
@paulstelian97  audio_stream functions internally deal with wrap around, so I assumed that it is ok for now.

File: src/include/sof/audio/audio_stream.h [hunk 1]
```diff
@@ -584,6 +584,83 @@ static inline void audio_stream_copy(const struct audio_stream *source,
...
+		ret = memcpy_s(snk, bytes_snk, src, bytes_copied);
+		assert(!ret);
+
+		bytes -= bytes_copied;
+		src = (char *)src + bytes_copied;
+		snk = (char *)snk + bytes_copied;
+
+		src = audio_stream_wrap(source, src);
+	}
+}
```
Comment (ktrzcinx) [inline src/include/sof/audio/audio_stream.h:624]: There's no need for while look - only source is circular buffer, let's look for my solution:

```
static void audio_stream_copy_to_buf(const struct audio_stream *source, void *lib_buff, size_t bytes)
{
	uint32_t head_size = MIN(bytes, audio_stream_bytes_without_wrap(source, source->r_ptr));
	uint32_t tail_size = bytes - head_size;

	memcpy(lib_buff, source->r_ptr, head_size);
	if (tail_size)
		memcpy((uint8_t *)lib_buff + head_size, source->addr, tail_size);
}
```

**Comment by paulstelian97** [comment]
> @paulstelian97 audio_stream functions internally deal with wrap around, so I assumed that it is ok for now.

With the wraparound they do (you won't overrun the buffer itself). But none of the functions you have used check the avail/free bytes. So you can read from an empty buffer/write in a full one. You won't get anything that Valgrind would complain about in the unit tests, that's true. But you can extract 1536 bytes from a 768 byte buffer (and you'd see the contents of that 768 byte buffer twice in the extracted data), for example.

**Comment by dbaluta** [comment]
@mrajwa I'm bringing this PR in order to discuss how we should copy the data from SOF component to library buffer.

If my understanding is correct, this code assumes:

https://github.com/thesofproject/sof/blob/8862fb1fac9e80c54f5833db7f08186336d63cf8/src/audio/codec_adapter/codec_adapter.c#L294

that data copied is not compressed, while this is not true for most of the decoders.

Other issue, is that after processing is done library output buffer size might not be equal with input buffer size.

**Comment by lgirdwood** [comment]
@dbaluta any update here ?

**Comment by dbaluta** [comment]
@lgirdwood this patch together with suggestions was picked up by @mrajwa in https://github.com/thesofproject/sof/pull/3361/commits/97a948276691d6e18e23949f52be03ceef306be9

@mrajwa, as I suggested in your commit lets move these copy functions in buffer.h and use them in your PR. I can do that, you only need to use it in your PR.

**Comment by lgirdwood** [comment]
@dbaluta ok, so does this mean we can close or are there other parts of this PR to merge ?

**Comment by dbaluta** [comment]
@lgirdwood we can close this after @mrajwa takes care of my comments for updating https://github.com/thesofproject/sof/commit/97a948276691d6e18e23949f52be03ceef306be9

**Pull request closed without merging**